Framework for copy-on-write visitors (WIP)

ChrisDodd opened the pull request:
This is still very much a work in progress (doesn't even build at the moment due to order-of-declaration issues), but is complete enough to show my thinking.

The idea here is address #4418 (excessive cloning in Transform/Modifier visitors) by doing lazy clones -- nodes are only cloned when they are modified (true copy on write), by giving the preorder/postorder methods "smart" pointers that manage the cloning when fields are actually modified.  This uses the ir generator to make special COWref reference objects for each IR class that understand the fields in the IR class and can be assigned to as if they were instances of the class, managing the cloning on demand.

Since we've been saying we're going to update to requiring C++20, this uses C++20 concepts in a number of places to simplify the overloading.

ChrisDodd committed:
Framework for copy-on-write visitors (WIP)

- COWptr/COWref objects that support cloning on demand to allow writes
- Lots of autogenerated code for IR classes to make things work
  transparently
- specialization for vectors/maps etc
- rename IR::MAU::Primitive -> IR::MAU::MauPrimitive to avoid name
  ambiguity with IR::Primitive

Signed-off-by: Chris Dodd <cdodd@nvidia.com>

asl commented:
@ChrisDodd Great! How does it handle children visiting? E.g. when a child is changed, how do we clone the parent? Especially when e.g. inside `Vector` or `IndexedVector`?

ChrisDodd committed:
Fix code to avoid being broken by optimizer

- don't try to create/copy temp COWref objects as they sometimes get
  prematurely destroyed -- instead reinterpret_cast the COWptr obj
- use std::launder to ensure the compiler does not try to eliminate
  things due to being the wrong type after a reinterpret_cast.

Signed-off-by: Chris Dodd <cdodd@nvidia.com>

ChrisDodd commented:
The container image build fails with errors like

    /p4c/ir/copy_on_write_ptr.h:34:1: error: 'concept' does not name a type
       34 | concept COWref = requires(T r) {
          | ^~~~~~~
    /p4c/ir/copy_on_write_ptr.h:34:1: note: 'concept' only available with '-fconcepts'
    /p4c/ir/copy_on_write_ptr.h:39:2: error: extra ';' [-Werror=pedantic]
       39 | };
          |  ^
    /p4c/ir/copy_on_write_ptr.h:123:5: error: 'requires' does not name a type
      123 |     requires std::derived_from<typename U::template COW_nested<T, COWfieldref<T, U, field>>,  
          |     ^~~~~~~~
    /p4c/ir/copy_on_write_ptr.h:128:2: error: extra ';' [-Werror=pedantic]
      128 | };
          |  ^
    /p4c/ir/copy_on_write_ptr.h:153:9: error: 'requires' does not name a type
      153 |         requires std::derived_from<U, T>
          |         ^~~~~~~~

which indicates that the compiler used does not support C++20.  So something is wrong with the compiler setup, since we currently require C++20

asl commented:
> which indicates that the compiler used does not support C++20. So something is wrong with the compiler setup, since we currently require C++20

Concepts are claimed to be supported since GCC 10. Container is with GCC 9.

fruffy commented:
> The container image build fails with errors like
> 
> ```
> /p4c/ir/copy_on_write_ptr.h:34:1: error: 'concept' does not name a type
>    34 | concept COWref = requires(T r) {
>       | ^~~~~~~
> /p4c/ir/copy_on_write_ptr.h:34:1: note: 'concept' only available with '-fconcepts'
> /p4c/ir/copy_on_write_ptr.h:39:2: error: extra ';' [-Werror=pedantic]
>    39 | };
>       |  ^
> /p4c/ir/copy_on_write_ptr.h:123:5: error: 'requires' does not name a type
>   123 |     requires std::derived_from<typename U::template COW_nested<T, COWfieldref<T, U, field>>,  
>       |     ^~~~~~~~
> /p4c/ir/copy_on_write_ptr.h:128:2: error: extra ';' [-Werror=pedantic]
>   128 | };
>       |  ^
> /p4c/ir/copy_on_write_ptr.h:153:9: error: 'requires' does not name a type
>   153 |         requires std::derived_from<U, T>
>       |         ^~~~~~~~
> ```
> 
> which indicates that the compiler used does not support C++20. So something is wrong with the compiler setup, since we currently require C++20

It;s possible the container image is too old, it uses Ubuntu 20.04. Trying to upgrade it right now at https://github.com/p4lang/third-party/

But the process is involved.
